cli: add command or flag to actively remove node from cluster

We need some kind of command or flag to actively remove a node from the cluster when there's no intention to bring that node back up.

At a minimum, this would:
- Ensure that the node not show up in the admin UI in the future
- Migrate replicas off the stores on that node in a clean way
- Prevent the stream of log entries that occur when you currently kill a node. For example:

```
I160421 11:32:01.362135 /Users/cdo/go/src/google.golang.org/grpc/clientconn.go:317  Failed to dial cdo-cockroach.local:26259: grpc: timed out trying to connect; please retry.
W160421 11:32:10.651793 ts/db.go:98  error writing time series data: failed to send RPC: too many errors encountered (1 of 1 total): rpc error: code = 4 desc = context deadline exceeded
```
- Shut down the CockroachDB instance on the node.
- Probably log an entry in the event log

This somewhat decomposes into a few more granular features:

* drain a node (i.e. let all leader leases expire and don't obtain new
ones, disable all queues, maybe prevent new SQL connections - but keep
making Raft progress) and wait until the slate is clean
  Used for clean shutdown (and pretty much always part of any type of
shutdown)
* drain data off a node (i.e. have all or some subset of ranges migrate
away, and wait until that has happened)
  Used for decommissioning, but could also be used to change hard-drives
cleanly (i.e. drain data off a store).
* drain for update/migration (see the `{F,Unf}reeze` proposal in #6166)
and finally
* decommission (permanently remove), which is "most of the above".

On Thu, Apr 21, 2016 at 11:43 AM Cuong Do <notifications@github.com> wrote:

> We need some kind of command or flag to actively remove a node from the
> cluster when there's no intention to bring that node back up.
>
> At a minimum, this would:
>
>    - Ensure that the node not show up in the admin UI in the future
>    - Migrate replicas off the stores on that node in a clean way
>    - Prevent the stream of log entries that occur when you currently kill
>    a node. For example:
>
> I160421 11:32:01.362135 /Users/cdo/go/src/google.golang.org/grpc/clientconn.go:317  Failed to dial cdo-cockroach.local:26259: grpc: timed out trying to connect; please retry.
> W160421 11:32:10.651793 ts/db.go:98  error writing time series data: failed to send RPC: too many errors encountered (1 of 1 total): rpc error: code = 4 desc = context deadline exceeded
>
>
>    - Shut down the CockroachDB instance on the node.
>    - Probably log an entry in the event log
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/issues/6198>
>
-- 

-- Tobias

--- a/cli/flags.go
+++ b/cli/flags.go
@@ -400,11 +400,6 @@ func initFlags(ctx *Context) {
 		f.Var(cacheSize, cliflags.CacheName, usageNoEnv(cliflags.CacheName))
 	}
 
-	{
-		f := exterminateCmd.Flags()
-		f.Var(&ctx.Stores, cliflags.StoreName, usageNoEnv(cliflags.StoreName))
-	}
-
 	for _, cmd := range certCmds {
 		f := cmd.Flags()
 		// Certificate flags.
@@ -418,7 +413,7 @@ func initFlags(ctx *Context) {
 	setUserCmd.Flags().StringVar(&password, cliflags.PasswordName, envutil.EnvOrDefaultString(cliflags.PasswordName, ""), usageEnv(cliflags.PasswordName))
 
 	clientCmds := []*cobra.Command{
-		sqlShellCmd, exterminateCmd, quitCmd, freezeClusterCmd, /* startCmd is covered above */
+		sqlShellCmd, quitCmd, freezeClusterCmd, /* startCmd is covered above */
 	}
 	clientCmds = append(clientCmds, kvCmds...)
 	clientCmds = append(clientCmds, rangeCmds...)
@@ -447,7 +442,7 @@ func initFlags(ctx *Context) {
 	}
 
 	// Commands that need the cockroach port.
-	simpleCmds := []*cobra.Command{exterminateCmd, freezeClusterCmd}
+	simpleCmds := []*cobra.Command{freezeClusterCmd}
 	simpleCmds = append(simpleCmds, kvCmds...)
 	simpleCmds = append(simpleCmds, rangeCmds...)
 	for _, cmd := range simpleCmds {

--- a/cli/start.go
+++ b/cli/start.go
@@ -37,7 +37,6 @@ import (
 	"github.com/cockroachdb/cockroach/rpc"
 	"github.com/cockroachdb/cockroach/security"
 	"github.com/cockroachdb/cockroach/server"
-	"github.com/cockroachdb/cockroach/storage/engine"
 	"github.com/cockroachdb/cockroach/util"
 	"github.com/cockroachdb/cockroach/util/hlc"
 	"github.com/cockroachdb/cockroach/util/log"
@@ -284,44 +283,6 @@ func rerunBackground() error {
 	cmd.Stdout = os.Stdout
 	cmd.Stderr = os.Stderr
 	return sdnotify.Exec(cmd)
-}
-
-// exterminateCmd command shuts down the node server and
-// destroys all data held by the node.
-var exterminateCmd = &cobra.Command{
-	Use:   "exterminate",
-	Short: "destroy all data held by the node",
-	Long: `
-First shuts down the system and then destroys all data held by the
-node, cycling through each store specified by --store flags.
-`,
-	SilenceUsage: true,
-	RunE:         runExterminate,
-}
-
-// runExterminate destroys the data held in the specified stores.
-func runExterminate(_ *cobra.Command, _ []string) error {
-	stopper := stop.NewStopper()
-	defer stopper.Stop()
-	if err := cliContext.InitStores(stopper); err != nil {
-		return util.Errorf("failed to initialize context: %s", err)
-	}
-
-	if err := runQuit(nil, nil); err != nil {
-		return util.Errorf("shutdown node error: %s", err)
-	}
-
-	// Exterminate all data held in specified stores.
-	for _, e := range cliContext.Engines {
-		if rocksdb, ok := e.(*engine.RocksDB); ok {
-			log.Infof("exterminating data from store %s", e)
-			if err := rocksdb.Destroy(); err != nil {
-				return util.Errorf("unable to destroy store %s: %s", e, err)
-			}
-		}
-	}
-	log.Infof("exterminated all data from stores %s", cliContext.Engines)
-	return nil
 }
 
 // quitCmd command shuts down the node server.

